### DISCUSSION TITLE
How can I toggle the default value of a multi-select choice

### DISCUSSION THREAD

**Discussion by dansiegel** [dotnet/templating#6489] [General]
I have a template which has 2 quick presets for whether to start from a blank or recommended state. All options then become optional to from the blank state or to disable from the recommended state. Based on this setting I dynamically adjust the default values through a combination of a generated switch default and a generated coalesce value to ensure that when a value has been selected by a user that we respect what the user provided rather than using the preset default.

```json
    "tests": {
      "displayName": "Tests",
      "type": "parameter",
      "datatype": "choice",
      "allowMultipleValues": true,
      "enableQuotelessLiterals": true,
      "choices": [
        {
          "choice": "unit",
          "displayName": "Unit Tests"
        },
        {
          "choice": "ui",
          "displayName": "UI Tests"
        }
      ]
    },
    "presetTestsDefault": {
      "type": "generated",
      "generator": "switch",
      "parameters": {
        "evaluator": "C++",
        "datatype": "string",
        "cases": [
          {
            "condition": "(preset == 'recommended')",
            "value": "unit|ui"
          },
          {
            "condition": "(preset == 'blank')",
            "value": ""
          }
        ]
      }
    },
    "testsEvaluator": {
      "type": "generated",
      "generator": "coalesce",
      "parameters": {
        "sourceVariableName": "tests",
        "fallbackVariableName": "presetTestsDefault"
      }
    },
```

The problem I am running into is supporting when someone chooses the recommended preset and they do not want any tests. If we do not rely on a dynamic default for example we could set the default to `unit|ui` and then when the template is generated you would simply do `dotnet new mytemplate -tests ""` and it will evaluate with no tests. However using the generated default and coalesce value when `tests == ""` it will fallback to the preset default. Can someone help provide some recommendations for how I can accomplish the goal while still allowing someone to choose the recommended preset but remove the tests if they want.

**Comment by YuliiaKovalova** [comment]
@JanKrivanek , could you help us with this one?

**Comment by GangWang01** [comment]
@dansiegel 
To resolve your problem, you can add one more choice `none` into the symbol `tests` that indicates no tests are selected. In this way `testsEvaluator` will not fallback to `presetTestsDefault` if `--tests none` is provided. Hope this can help.

**Comment by JanKrivanek** [comment]
IIRC your scenario should be supported - explicitly passed empty should be respect by coallesce (per https://github.com/dotnet/templating/issues/4737 and https://github.com/dotnet/templating/issues/5776).

If that doesn't work than it sounds as a regression.

Gangs workaround feels as an easy way out of this though

**Reply by GangWang01** [reply]
> IIRC your scenario should be supported - explicitly passed empty should be respect by coallesce (per #4737 and #5776).
> 
> If that doesn't work than it sounds as a regression.

Currently coalesce treats empty value for string same as null. Empty string as user input is not respected according to https://github.com/dotnet/templating/blob/ac309f4eaa3d11574afac8cfd94e84ebb698f4eb/src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Macros/CoalesceMacro.cs#LL39C1-L43C18. It needs a little fix.

**Comment by nickrandolph** [comment]
@JanKrivanek unfortunately the solution proposed by @GangWang01 is the way we've gone but is less than intuitive for end users. Instead of just specifying "" as the list of test options, they have to specify "none"  and what if they specify "none,ui" which is possible since it's multi-select.
My concern with the apparent regression in the coallesce is that even when this is fixed, how are we supposed to rely on it, as we've got no way to verify that the user has got that fix. The upshot is that we're going to be stuck with the workaround until we force users to be on a later version of VS (we currently target 17.4 as minimum)

**Comment by GangWang01** [comment]
@nickrandolph 
The problem you hit is fixed and become available in .NET 8 Preview 6 or .NET 7 (sdk 7.0.4xx). 
Coalesce will respect empty string value for multi-choice parameter and not fallback.
